File: services/omai-control-plane/internal/transport/connectapi/interactions.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	platformv1 "github.com/omai/backend/gen/go/omai/platform/v1"
	"github.com/omai/backend/internal/application"
	"github.com/omai/backend/internal/domain"
)

type PermissionService struct {
	Interactions *application.Interactions
}

func (s *PermissionService) ListPermissions(ctx context.Context, request *connect.Request[platformv1.ListPermissionsRequest]) (*connect.Response[platformv1.ListPermissionsResponse], error) {
	principal, err := requirePrincipal(ctx)
	if err != nil {
		return nil, err
	}
	permissions, next, err := s.Interactions.ListPermissions(
		ctx, principal, request.Msg.GetProjectId(), request.Msg.GetSessionId(), int(request.Msg.GetPageSize()), request.Msg.GetPageToken(),
	)
	if err != nil {
		return nil, connectError(err)
	}
	response := &platformv1.ListPermissionsResponse{NextPageToken: next}
	for _, permission := range permissions {
		response.Permissions = append(response.Permissions, permissionV1(permission))
	}
	return connect.NewResponse(response), nil
}

func (s *PermissionService) RespondPermission(ctx context.Context, request *connect.Request[platformv1.RespondPermissionRequest]) (*connect.Response[platformv1.RespondPermissionResponse], error) {
	principal, err := requirePrincipal(ctx)
	if err != nil {
		return nil, err
	}
	permission, err := s.Interactions.RespondPermission(
		ctx, principal, request.Msg.GetSessionId(), request.Msg.GetPermissionId(), permissionDecision(request.Msg.GetDecision()),
	)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&platformv1.RespondPermissionResponse{Permission: permissionV1(permission)}), nil
}

type QuestionService struct {
	Interactions *application.Interactions
}

func (s *QuestionService) ListQuestions(ctx context.Context, request *connect.Request[platformv1.ListQuestionsRequest]) (*connect.Response[platformv1.ListQuestionsResponse], error) {
	principal, err := requirePrincipal(ctx)
	if err != nil {
		return nil, err
	}
	questions, next, err := s.Interactions.ListQuestions(
		ctx, principal, request.Msg.GetProjectId(), request.Msg.GetSessionId(), int(request.Msg.GetPageSize()), request.Msg.GetPageToken(),
	)
	if err != nil {
		return nil, connectError(err)
	}
	response := &platformv1.ListQuestionsResponse{NextPageToken: next}
	for _, question := range questions {
		response.Questions = append(response.Questions, questionV1(question))
	}
	return connect.NewResponse(response), nil
}

func (s *QuestionService) ReplyQuestion(ctx context.Context, request *connect.Request[platformv1.ReplyQuestionRequest]) (*connect.Response[platformv1.ReplyQuestionResponse], error) {
	principal, err := requirePrincipal(ctx)
	if err != nil {
		return nil, err
	}
	answers := make([][]string, 0, len(request.Msg.GetAnswers()))
	for _, answer := range request.Msg.GetAnswers() {
		answers = append(answers, append([]string(nil), answer.GetValues()...))
	}
	question, err := s.Interactions.ReplyQuestion(ctx, principal, request.Msg.GetSessionId(), request.Msg.GetQuestionId(), answers, false)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&platformv1.ReplyQuestionResponse{Question: questionV1(question)}), nil
}

func (s *QuestionService) RejectQuestion(ctx context.Context, request *connect.Request[platformv1.RejectQuestionRequest]) (*connect.Response[platformv1.RejectQuestionResponse], error) {
	principal, err := requirePrincipal(ctx)
	if err != nil {
		return nil, err
	}
	question, err := s.Interactions.ReplyQuestion(ctx, principal, request.Msg.GetSessionId(), request.Msg.GetQuestionId(), nil, true)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&platformv1.RejectQuestionResponse{Question: questionV1(question)}), nil
}

func permissionV1(value domain.PermissionRequest) *platformv1.Permission {
	result := &platformv1.Permission{
		Id: value.ID, SessionId: value.SessionID, ProjectId: value.ProjectID,
		Permission: value.Permission, Patterns: append([]string(nil), value.Patterns...),
		MetadataJson: append([]byte(nil), value.MetadataJSON...), Always: append([]string(nil), value.Always...),
		Decision: permissionDecisionV1(value.Decision), CreatedUnixMillis: value.CreatedAt.UnixMilli(),
	}
	if !value.ResolvedAt.IsZero() {
		result.ResolvedUnixMillis = value.ResolvedAt.UnixMilli()
	}
	if value.Tool != nil {
		result.Tool = &platformv1.InteractionToolReference{MessageId: value.Tool.MessageID, CallId: value.Tool.CallID}
	}
	return result
}

func questionV1(value domain.QuestionRequest) *platformv1.QuestionRequestResource {
	result := &platformv1.QuestionRequestResource{
		Id: value.ID, SessionId: value.SessionID, ProjectId: value.ProjectID,
		Rejected: value.Rejected, CreatedUnixMillis: value.CreatedAt.UnixMilli(),
	}
	if !value.ResolvedAt.IsZero() {
		result.ResolvedUnixMillis = value.ResolvedAt.UnixMilli()
	}
	if value.Tool != nil {
		result.Tool = &platformv1.InteractionToolReference{MessageId: value.Tool.MessageID, CallId: value.Tool.CallID}
	}
	for _, question := range value.Questions {
		item := &platformv1.Question{
			Question: question.Question, Header: question.Header, Multiple: question.Multiple, Custom: question.Custom,
		}
		for _, option := range question.Options {
			item.Options = append(item.Options, &platformv1.QuestionOption{Label: option.Label, Description: option.Description})
		}
		result.Questions = append(result.Questions, item)
	}
	for _, answer := range value.Answers {
		result.Answers = append(result.Answers, &platformv1.QuestionAnswer{Values: append([]string(nil), answer...)})
	}
	return result
}

func permissionDecision(value platformv1.PermissionDecision) domain.PermissionDecision {
	switch value {
	case platformv1.PermissionDecision_PERMISSION_DECISION_ONCE:
		return domain.PermissionDecisionOnce
	case platformv1.PermissionDecision_PERMISSION_DECISION_ALWAYS:
		return domain.PermissionDecisionAlways
	case platformv1.PermissionDecision_PERMISSION_DECISION_REJECT:
		return domain.PermissionDecisionReject
	default:
		return ""
	}
}

func permissionDecisionV1(value domain.PermissionDecision) platformv1.PermissionDecision {
	switch value {
	case domain.PermissionDecisionOnce:
		return platformv1.PermissionDecision_PERMISSION_DECISION_ONCE
	case domain.PermissionDecisionAlways:
		return platformv1.PermissionDecision_PERMISSION_DECISION_ALWAYS
	case domain.PermissionDecisionReject:
		return platformv1.PermissionDecision_PERMISSION_DECISION_REJECT
	default:
		return platformv1.PermissionDecision_PERMISSION_DECISION_UNSPECIFIED
	}
}

File: services/omai-control-plane/sdk/typescript/src/interactions.ts
import { Code, type CallOptions, type Client } from "@connectrpc/connect"
import {
  PermissionDecision,
  type Permission,
  type PermissionService,
  type QuestionRequestResource,
  type QuestionService,
} from "./gen/omai/platform/v1/platform_pb.js"
import { decodeJsonBytes } from "./json.js"
import { OMAIError } from "./errors.js"

export interface OMAIToolReference {
  readonly messageId: string
  readonly callId: string
}

export interface OMAIPermission {
  readonly id: string
  readonly sessionId: string
  readonly projectId: string
  readonly permission: string
  readonly patterns: readonly string[]
  readonly metadata: Readonly<Record<string, unknown>>
  readonly always: readonly string[]
  readonly tool?: OMAIToolReference
  readonly decision?: "once" | "always" | "reject"
  readonly createdUnixMillis: bigint
  readonly resolvedUnixMillis?: bigint
}

export interface OMAIQuestionOption {
  readonly label: string
  readonly description: string
}

export interface OMAIQuestion {
  readonly question: string
  readonly header: string
  readonly options: readonly OMAIQuestionOption[]
  readonly multiple: boolean
  readonly custom: boolean
}

export interface OMAIQuestionRequest {
  readonly id: string
  readonly sessionId: string
  readonly projectId: string
  readonly questions: readonly OMAIQuestion[]
  readonly tool?: OMAIToolReference
  readonly answers: readonly (readonly string[])[]
  readonly rejected: boolean
  readonly createdUnixMillis: bigint
  readonly resolvedUnixMillis?: bigint
}

export interface OMAIInteractionFilter {
  readonly projectId?: string
  readonly sessionId?: string
}

export interface OMAIPermissions {
  list(filter: OMAIInteractionFilter, options?: CallOptions): Promise<readonly OMAIPermission[]>
  respond(
    sessionId: string,
    permissionId: string,
    decision: "once" | "always" | "reject",
    options?: CallOptions,
  ): Promise<OMAIPermission>
}

export interface OMAIQuestions {
  list(filter: OMAIInteractionFilter, options?: CallOptions): Promise<readonly OMAIQuestionRequest[]>
  reply(
    sessionId: string,
    questionId: string,
    answers: readonly (readonly string[])[],
    options?: CallOptions,
  ): Promise<OMAIQuestionRequest>
  reject(sessionId: string, questionId: string, options?: CallOptions): Promise<OMAIQuestionRequest>
}

export function createPermissions(client: Client<typeof PermissionService>): OMAIPermissions {
  return Object.freeze({
    async list(filter: OMAIInteractionFilter, options?: CallOptions) {
      const checked = checkedFilter(filter)
      const result: OMAIPermission[] = []
      let pageToken = ""
      for (let page = 0; page < 10_000; page++) {
        const response = await client.listPermissions({ ...checked, pageSize: 200, pageToken }, options)
        result.push(...response.permissions.map(permissionView))
        if (!response.nextPageToken) return Object.freeze(result)
        if (response.nextPageToken === pageToken) throw repeatedPageToken("permission")
        pageToken = response.nextPageToken
      }
      throw pageLimit("permission")
    },
    async respond(
      sessionId: string,
      permissionId: string,
      decision: "once" | "always" | "reject",
      options?: CallOptions,
    ) {
      const response = await client.respondPermission(
        {
          sessionId: checkedID(sessionId, "session"),
          permissionId: checkedID(permissionId, "permission"),
          decision: permissionDecision(decision),
        },
        options,
      )
      return permissionView(required(response.permission, "permission"))
    },
  })
}

export function createQuestions(client: Client<typeof QuestionService>): OMAIQuestions {
  return Object.freeze({
    async list(filter: OMAIInteractionFilter, options?: CallOptions) {
      const checked = checkedFilter(filter)
      const result: OMAIQuestionRequest[] = []
      let pageToken = ""
      for (let page = 0; page < 10_000; page++) {
        const response = await client.listQuestions({ ...checked, pageSize: 200, pageToken }, options)
        result.push(...response.questions.map(questionView))
        if (!response.nextPageToken) return Object.freeze(result)
        if (response.nextPageToken === pageToken) throw repeatedPageToken("question")
        pageToken = response.nextPageToken
      }
      throw pageLimit("question")
    },
    async reply(sessionId: string, questionId: string, answers: readonly (readonly string[])[], options?: CallOptions) {
      if (answers.length < 1 || answers.length > 32) throw new RangeError("Question answers are invalid")
      const response = await client.replyQuestion(
        {
          sessionId: checkedID(sessionId, "session"),
          questionId: checkedID(questionId, "question"),
          answers: answers.map((values) => ({ values: checkedAnswers(values) })),
        },
        options,
      )
      return questionView(required(response.question, "question"))
    },
    async reject(sessionId: string, questionId: string, options?: CallOptions) {
      const response = await client.rejectQuestion(
        { sessionId: checkedID(sessionId, "session"), questionId: checkedID(questionId, "question") },
        options,
      )
      return questionView(required(response.question, "question"))
    },
  })
}

export function permissionView(value: Permission): OMAIPermission {
  const metadata = value.metadataJson.byteLength === 0 ? {} : decodeJsonBytes(value.metadataJson, 64 * 1024)
  if (typeof metadata !== "object" || metadata === null || Array.isArray(metadata)) {
    throw new OMAIError("OMAI returned invalid permission metadata", { code: Code.DataLoss })
  }
  const tool = value.tool
  return Object.freeze({
    id: value.id,
    sessionId: value.sessionId,
    projectId: value.projectId,
    permission: value.permission,
    patterns: Object.freeze([...value.patterns]),
    metadata: Object.freeze(metadata as Record<string, unknown>),
    always: Object.freeze([...value.always]),
    ...(tool === undefined ? {} : { tool: Object.freeze({ messageId: tool.messageId, callId: tool.callId }) }),
    ...(value.decision === PermissionDecision.UNSPECIFIED ? {} : { decision: permissionDecisionView(value.decision) }),
    createdUnixMillis: value.createdUnixMillis,
    ...(value.resolvedUnixMillis === 0n ? {} : { resolvedUnixMillis: value.resolvedUnixMillis }),
  })
}

export function questionView(value: QuestionRequestResource): OMAIQuestionRequest {
  const tool = value.tool
  return Object.freeze({
    id: value.id,
    sessionId: value.sessionId,
    projectId: value.projectId,
    questions: Object.freeze(
      value.questions.map((question) =>
        Object.freeze({
          question: question.question,
          header: question.header,
          options: Object.freeze(
            question.options.map((option) => Object.freeze({ label: option.label, description: option.description })),
          ),
          multiple: question.multiple,
          custom: question.custom,
        }),
      ),
    ),
    ...(tool === undefined ? {} : { tool: Object.freeze({ messageId: tool.messageId, callId: tool.callId }) }),
    answers: Object.freeze(value.answers.map((answer) => Object.freeze([...answer.values]))),
    rejected: value.rejected,
    createdUnixMillis: value.createdUnixMillis,
    ...(value.resolvedUnixMillis === 0n ? {} : { resolvedUnixMillis: value.resolvedUnixMillis }),
  })
}

function checkedFilter(filter: OMAIInteractionFilter): { projectId: string; sessionId: string } {
  if (filter.projectId === undefined && filter.sessionId === undefined) {
    throw new TypeError("Interaction filter requires a projectId or sessionId")
  }
  return {
    projectId: filter.projectId === undefined ? "" : checkedID(filter.projectId, "project"),
    sessionId: filter.sessionId === undefined ? "" : checkedID(filter.sessionId, "session"),
  }
}

function checkedAnswers(values: readonly string[]): string[] {
  if (values.length < 1 || values.length > 64) throw new RangeError("Question answer set is invalid")
  return values.map((value) => {
    const trimmed = value.trim()
    if (!trimmed || trimmed.length > 4096 || /\0/u.test(trimmed)) throw new TypeError("Question answer is invalid")
    return trimmed
  })
}

function checkedID(value: string, label: string): string {
  const trimmed = value.trim()
  if (!trimmed || trimmed.length > 512 || /[\0\r\n]/u.test(trimmed)) throw new TypeError(`Invalid OMAI ${label} ID`)
  return trimmed
}

function permissionDecision(value: "once" | "always" | "reject"): PermissionDecision {
  if (value === "once") return PermissionDecision.ONCE
  if (value === "always") return PermissionDecision.ALWAYS
  return PermissionDecision.REJECT
}

export function permissionDecisionView(value: PermissionDecision): "once" | "always" | "reject" {
  if (value === PermissionDecision.ONCE) return "once"
  if (value === PermissionDecision.ALWAYS) return "always"
  if (value === PermissionDecision.REJECT) return "reject"
  throw new OMAIError("OMAI returned an invalid permission decision", { code: Code.DataLoss })
}

function required<T>(value: T | undefined, label: string): T {
  if (value === undefined) throw new OMAIError(`OMAI returned no ${label}`, { code: Code.Internal })
  return value
}

function repeatedPageToken(label: string) {
  return new OMAIError(`OMAI returned a repeated ${label} page token`, { code: Code.DataLoss })
}

function pageLimit(label: string) {
  return new OMAIError(`OMAI ${label} pagination exceeded its safety limit`, { code: Code.ResourceExhausted })
}
